### PR TITLE
Add "exclude" capabilities to filters.py 

### PR DESCRIPTION
**Pull request by lbeaufort**
Add capability to search for results not matching a certain value

Update `filters.py` : `filter_match`, `filter_multi`, `filter_fulltext` to use `not_in` functionality to exclude results with a certain value when passed `parameter=-value`. I also check for when API is passed "key=A and key!=B" and handled that behavior by splitting out the arguments

- [x] Use list comprehension in `filter_multi`?
- [x] Add automated tests
- [x] Use [instance](https://github.com/18F/openFEC/pull/2937#discussion_r169325116)
- [x] update all filter functions: `filter_fulltext`
- [x] Add combo exclude/include test for `filter_multi`

___
Filters that didn't get updated with exclude capabilities:
`filter_range` exclude - possibly for the future, this hasn't been requested.
`filter_contributor_type`? `filter_election` - not needed

### CODE REVIEW

File: webservices/filters.py [hunk 1]
```diff
@@ -5,18 +5,42 @@
 from webservices.common import models
 
 
+def is_exclude_arg(arg):
+    # Handle string and int excludes
+    return str(arg).startswith('-')
```
Comment (ccostino) [inline webservices/filters.py]: I think this is an appropriate abstraction here - if we ever change how we want to handle URL parameters for filtering, particularly around exclusion, we just have to worry about this spot. :-)

File: webservices/filters.py [hunk 2]
```diff
@@ -5,18 +5,42 @@
...
+
+
+def parse_exclude_arg(arg):
+    # Integers will come in as negative and strings will start with -
+    if type(arg) is int:
+        return abs(arg)
+    else:
+        return arg[1:]
```
Comment (ccostino) [inline webservices/filters.py]: Nicely done, accounting for different parameter types!

File: webservices/filters.py [hunk 3]
```diff
@@ -5,18 +5,42 @@
...
 
 
+def is_exclude_arg(arg):
+    # Handle string and int excludes
+    return str(arg).startswith('-')
+
+
+def parse_exclude_arg(arg):
+    # Integers will come in as negative and strings will start with -
+    if type(arg) is int:
```
Comment (ccostino) [inline webservices/filters.py]: You might want to consider using [`isinstance`](https://docs.python.org/3/library/functions.html#isinstance) here instead (here's a [detailed explanation why on Stack Overflow](https://stackoverflow.com/a/1549854)), this way if `WebArgs` ever changes how parameters of an `int` type are handled by subclassing, we'll be covered. :-)

As an example, this line would become:

```python
if isinstance(arg, int):
```

File: webservices/filters.py [hunk 4]
```diff
@@ -5,18 +5,42 @@
...
+
+
 def filter_match(query, kwargs, fields):
     for key, column in fields:
         if kwargs.get(key) is not None:
-            query = query.filter(column == kwargs[key])
+            if is_exclude_arg(kwargs[key]):
+                query = query.filter(column != parse_exclude_arg(kwargs[key]))
+            else:
+                query = query.filter(column == kwargs[key])
```
Comment (ccostino) [inline webservices/filters.py]: Exactly what I had in mind, nice!

File: webservices/filters.py [hunk 5]
```diff
@@ -5,18 +5,42 @@
...
     for key, column in fields:
         if kwargs.get(key):
-            query = query.filter(column.in_(kwargs[key]))
+            # handle combination exclude/include lists
+            exclude_list = [parse_exclude_arg(value) for value in kwargs[key] if is_exclude_arg(value)]
+            include_list = [value for value in kwargs[key] if not is_exclude_arg(value)]
+            if exclude_list:
+                query = query.filter(column.notin_(exclude_list))
+            if include_list:
+                query = query.filter(column.in_(include_list))
```
Comment (ccostino) [inline webservices/filters.py]: Looks great so far!

File: webservices/filters.py [hunk 6]
```diff
@@ -25,23 +49,35 @@ def filter_range(query, kwargs, fields):
...
     for key, column in fields:
         if kwargs.get(key):
-            filters = [
-                column.match(utils.parse_fulltext(value))
-                for value in kwargs[key]
-            ]
-            query = query.filter(sa.or_(*filters))
+            exclude_list = [parse_exclude_arg(value) for value in kwargs[key] if is_exclude_arg(value)]
+            include_list = [value for value in kwargs[key] if not is_exclude_arg(value)]
+            if exclude_list:
```
Comment (lbeaufort) [inline webservices/filters.py]: @ccostino and @LindsayYoung - the `filter_fulltext` changes are the part I feel could use the most scrutiny.

Comment (ccostino) [reply webservices/filters.py]: I am mulling this over in my head and this all looks right to me.  The spot I'm not 100% on (but would be easy to check) is that the query gets constructed correctly given the mix of `and` and `or`.

I think this is constructed correctly, where things that are excluded are all `and` within the query, and things to be included are all combined with `or` (in other words:  a value is not `x` and `y` and `z`, but is `a` or `b` or `c`).  I just want to be sure. :-)

Comment (lbeaufort) [reply webservices/filters.py]: Thanks @ccostino! I agree - when I checked the queries I was a bit surprised not to see parentheses in there. I updated the issue to do lots more of manual testing, especially for combo include/exclude for fulltext filters. I've tested a fair amount so far but I think I might take an inventory of which endpoints use `match`, `multi` and `fulltext` to see if I can do some more extensive/comprehensive manual tests.

**Pull request merged**